File: src/main/java/ez/spring/vertx/web/verticle/HttpServerVerticle.java
package ez.spring.vertx.web.verticle;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;

import ez.spring.vertx.web.route.EzRouterUtil;
import ez.spring.vertx.web.route.RouteProps;
import io.vertx.core.AbstractVerticle;
import io.vertx.core.Promise;
import io.vertx.core.http.HttpServer;
import io.vertx.core.http.HttpServerOptions;
import io.vertx.core.json.Json;
import io.vertx.ext.web.Router;

@SuppressWarnings("WeakerAccess")
public class HttpServerVerticle extends AbstractVerticle {
  private static final Logger log = LoggerFactory.getLogger(HttpServerVerticle.class);
  private final HttpServerOptions options;
  private final List<RouteProps> routes;

  public HttpServerVerticle(HttpServerOptions options, List<RouteProps> routes) {
    this.options = options;
    this.routes = routes;
  }

  public List<RouteProps> getRoutes() {
    return routes;
  }

  @Override
  public void start(Promise<Void> startPromise) {
    HttpServer httpServer = vertx.createHttpServer(options);
    Router router = EzRouterUtil.setup(vertx, getRoutes());
    httpServer.requestHandler(router).listen(it -> {
      if (it.succeeded()) {
        log.info("using httpServer at {}:{}", options.getHost(), it.result().actualPort());
        startPromise.complete();
      } else {
        log.error("httpServer not available! options: {}", Json.encodePrettily(options));
        startPromise.fail(it.cause());
      }
    });
  }
}

File: src/main/java/ez/spring/vertx/web/verticle/HttpServerVerticle.kt
package ez.spring.vertx.web.verticle

import ez.spring.vertx.web.route.EzRouter
import ez.spring.vertx.web.route.RouteProps
import io.vertx.core.http.HttpServer
import io.vertx.core.http.HttpServerOptions
import io.vertx.core.json.Json
import io.vertx.kotlin.core.http.listenAwait
import io.vertx.kotlin.coroutines.CoroutineVerticle
import org.slf4j.Logger
import org.slf4j.LoggerFactory

class HttpServerVerticle(private val httpServerOptions: HttpServerOptions) : CoroutineVerticle() {
    private var ezRouter: EzRouter? = null
    private var routes: List<RouteProps> = emptyList()
    private var httpServer: HttpServer? = null

    fun setRoutes(routes: List<RouteProps>): HttpServerVerticle {
        this.routes = routes
        return this
    }

    override suspend fun start() {
        httpServer = vertx.createHttpServer(httpServerOptions)!!.apply {
            ezRouter = EzRouter.router(routes)
            try {
                requestHandler(ezRouter).listenAwait()
                log.info("using httpServer at {}:{}", httpServerOptions.host, actualPort())
            } catch (e: Throwable) {
                log.error("httpServer not available! options: {}", Json.encodePrettily(httpServerOptions))
                throw e
            }
        }
    }

    override suspend fun stop() {
        httpServer?.let {
            log.info("disconnect from httpServer at {}", it.actualPort())
            it.close()
        }
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(HttpServerVerticle::class.java)
    }
}